model/validation.go

```diff
@@ -204,16 +204,367 @@ func (t Tags) Validate() error {
+			errs = multierror.Append(errs, fmt.Errorf("Invalid tag key: %q", k))
+		}
+		if !tagRegexp.MatchString(v) {
+			errs = multierror.Append(errs, fmt.Errorf("Invalid tag value: %q", v))
+		}
+	}
+	return errs
+}
+
+func validateFQDN(fqdn string) error {
+	if len(fqdn) > 255 {
+		return fmt.Errorf("domain name %q too long (max 255)", fqdn)
+	}
+	if len(fqdn) == 0 {
+		return fmt.Errorf("empty domain name not allowed")
+	}
+
+	for _, label := range strings.Split(fqdn, ".") {
+		if !IsDNS1123Label(label) {
+			return fmt.Errorf("domain name %q invalid (label %q invalid)", fqdn, label)
+		}
+	}
+
+	return nil
+}
+
+// ValidateMatchCondition validates a Match Condition
+func ValidateMatchCondition(mc *proxyconfig.MatchCondition) error {
+	var retVal error
+
+	if mc.Source != "" {
+		if err := validateFQDN(mc.Source); err != nil {
+			retVal = multierror.Append(retVal, err)
+		}
+	}
+
+	if err := Validate(mc.SourceTags); err != nil {
+		retVal = multierror.Append(retVal, err)
+	}
+
+	if mc.GetTcp() != nil {
+		if err := ValidateL4MatchAttributes(mc.GetTcp()); err != nil {
+			retVal = multierror.Append(retVal, err)
+		}
+	}
+
+	if mc.GetUdp() != nil {
+		if err := ValidateL4MatchAttributes(mc.GetUdp()); err != nil {
+			retVal = multierror.Append(retVal, err)
+		}
+	}
+
+	// We do not (yet) validate http_headers.
+
+	return retVal
+}
+
+// ValidateL4MatchAttributes validates L4 Match Attributes
+func ValidateL4MatchAttributes(ma *proxyconfig.L4MatchAttributes) error {
+	var retVal error
+
+	if ma.SourceSubnet != nil {
+		for _, subnet := range ma.SourceSubnet {
+			if err := validateSubnet(subnet); err != nil {
+				retVal = multierror.Append(retVal, err)
+			}
+		}
+	}
+
+	if ma.DestinationSubnet != nil {
+		for _, subnet := range ma.DestinationSubnet {
+			if err := validateSubnet(subnet); err != nil {
+				retVal = multierror.Append(retVal, err)
+			}
+		}
+	}
+
+	return retVal
+}
+
+func validatePercent(err error, val int32, label string) error {
+	if val > 100 {
+		err = multierror.Append(err, fmt.Errorf("%v must not exceed 100", label))
+	}
+	if val < 0 {
+		err = multierror.Append(err, fmt.Errorf("%v must be in range 0..100", label))
+	}
+	return err
+}
+
+func validateFloatPercent(err error, val float32, label string) error {
+	if val > 100.0 {
+		err = multierror.Append(err, fmt.Errorf("%v must not exceed 100", label))
+	}
+	if val < 0.0 {
+		err = multierror.Append(err, fmt.Errorf("%v must be in range 0..100", label))
+	}
+	return err
+}
+
+// ValidateDestinationWeight validates DestinationWeight
+func ValidateDestinationWeight(dw *proxyconfig.DestinationWeight) error {
+	var retVal error
+
+	if dw.Destination != "" {
+		if err := validateFQDN(dw.Destination); err != nil {
+			retVal = multierror.Append(retVal, err)
+		}
+	}
+
+	// We do not validate tags because they have no explicit rules
+	if err := Validate(dw.Tags); err != nil {
+		retVal = multierror.Append(retVal, err)
+	}
+
+	retVal = validatePercent(retVal, dw.Weight, "weight")
+
```

Comment:
Two things.  Most importantly, this code currently only has access to the current POST or PUT and does not have the ability to globally sum all of the weights.  There might be another weight already there that brings this into compliance.

Second I have asked you in personal and via Slack why the sum of weights needs to be equal to 100 and what it means if it isn't and you have not yet answered.  Suppose I have three services and want to divide traffic equally.  Why must I have weights 33, 33, and 34 instead of 33, 33, and 33?  What happens if I delete a rule and then post a new rule and for a brief time the weight is only 66?  Please explain explain how these cases are intended to work.